[CHERRY-PICK] Move a few ARM based libraries to MdePkg/MdeModulePkg

kuqin12 opened the pull request:
## Description

This was a side effect when upstreaming some changes to EDK2. The upstream PR is here:
https://github.com/tianocore/edk2/pull/10929

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [ ] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [x] Backport to release branch?

## How This Was Tested

This was tested on QEMU SBSA and booted to UEFI shell.

## Integration Instructions

The following libraries will move:

| Old Path | New Path |
|-|-|
| ArmPkg/Library/ArmSvcLib/ArmSvcLib.inf | MdePkg/Library/ArmSvcLib/ArmSvcLib.inf |
| ArmPkg/Library/ArmSmcLib/ArmSmcLib.inf | MdePkg/Library/ArmSmcLib/ArmSmcLib.inf |
| ArmPkg/Library/ArmFfaLib/ArmFfaPeiLib.inf | MdeModulePkg/Library/ArmFfaLib/ArmFfaPeiLib.inf |
| ArmPkg/Library/ArmFfaLib/ArmFfaDxeLib.inf | MdeModulePkg/Library/ArmFfaLib/ArmFfaDxeLib.inf |
| ArmPkg/Library/ArmSmcLibNull/ArmSmcLibNull.inf | MdePkg/Library/ArmSmcLibNull/ArmSmcLibNull.inf |

kuqin12 committed:
ArmPkg: ArmSmcLib: Move ArmSmcLib to MdePkg

As an implementation of an industry specification-defined interface,
ArmSmcLib is more appropriately placed in MdePkg.

This update relocates both the header definition and the implementation
of ArmSmcLib to MdePkg, thereby supporting the monitor call primitives
for all other packages that depend on this interface.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 86099b948b6318b839bdbb41acad7211fadf5db8)

kuqin12 committed:
ArmPkg: ArmSmcLibNull: Move ArmSmcLibNull to MdePkg

ArmSmcLibNull is a null implementation of an industry specification-
defined interface, which is defined in MdePkg.

This update relocates the null implementation of ArmSmcLib to MdePkg,
thereby supporting the monitor call primitives for ARM platforms that
does not have SMC capabilities.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 211bb9b8395b24edf86a03933a545968599328eb)

kuqin12 committed:
ArmPkg: ArmStdSmc.h: Move to MdePkg

ArmStdSmc.h comprises a set of definitions established by an industry
standard specification:
https://developer.arm.com/documentation/den0028/latest/

This update relocates the header definition from ArmPkg to MdePkg to
maintain consistency within the EDK2 repo layout.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 9bcb2e41f7f761c2a43a67c4ef3403de80094621)

kuqin12 committed:
ArmPkg: FF-A headers: Move to MdePkg

ArmFfaBootInfo.h, ArmFfaPartInfo.h and ArmFfaSvc.h comprise a set of
definitions established by industry standard specifications:
https://developer.arm.com/documentation/den0077/latest

This update relocates the header definitions from ArmPkg to MdePkg to
maintain consistency within the EDK2 layout.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 014689399fedc93908bca3909c4d55d6a0ca306c)

kuqin12 committed:
ArmPkg: ArmFfaLib.h: Move to MdePkg

As a set of definitions from an industry specification-defined interface,
ArmFfaLib.h is more appropriately placed in MdePkg.

This update relocates both the header definition of ArmFfaLib to MdePkg,
thereby supporting the monitor call primitives for all other packages
that depend on this interface.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 937ffef8ed4bf20504fdc17806a36d1553c72619)

kuqin12 committed:
ArmPkg: ArmSvcLib: Move ArmSvcLib to MdePkg

As an implementation of an industry specification-defined interface,
ArmSvcLib is more appropriately placed in MdePkg.

This update relocates both the header definition and the implementation
of ArmSvcLib to MdePkg, thereby supporting the supervisor call primitives
for all other packages that depend on this interface.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 541c46a70550d948854470717668a30f0212296f)

kuqin12 committed:
ArmPkg: ArmFfaLib: Move ArmFfaLib implementation to MdeModulePkg

ArmFfaLib is an implementation of an industry specification-defined
interface with UEFI specific method of handling Rx/Tx buffer sharing
across multiple boot phases, which is more appropriately placed in
MdeModulePkg.

This update relocates the implementation of ArmFfaLib to MdeModulePkg,
thereby supporting the FFA call primitives for all other packages that
depend on this interface.

Continuous-integration-options: PatchCheck.ignore-multi-package
Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit be03ceb1176d71f2adc7634fb0f68491e743e699)

kuqin12 commented:
Pending PR: https://github.com/microsoft/mu_basecore/pull/1413

apop5 commented:
Would it be possible to update Readme.rst with these breaking changes?

https://github.com/microsoft/mu_silicon_arm_tiano/blob/dev/202502/ReadMe.rst

It will help to be able to point others to this document.

kuqin12 committed:
ArmPlatformPkg: ArmPlatformPkg.dsc: Fixing build due to library moves

Some ARM related libraries and headers are moved to MdeModulePkg and
MdePkg. This change updates the consumer paths to reflect the changes.

Signed-off-by: Kun Qin <kun.qin@microsoft.com>

kuqin12 committed:
ReadMe: Update breaking change notes for this repo

This change remarks the breaking changes brought in from EDK2.

Specifically, the following libraries will move:

| Old Path | New Path |
|-|-|
| ArmPkg/Library/ArmSvcLib/ArmSvcLib.inf | MdePkg/Library/ArmSvcLib/ArmSvcLib.inf |
| ArmPkg/Library/ArmSmcLib/ArmSmcLib.inf | MdePkg/Library/ArmSmcLib/ArmSmcLib.inf |
| ArmPkg/Library/ArmFfaLib/ArmFfaPeiLib.inf | MdeModulePkg/Library/ArmFfaLib/ArmFfaPeiLib.inf |
| ArmPkg/Library/ArmFfaLib/ArmFfaDxeLib.inf | MdeModulePkg/Library/ArmFfaLib/ArmFfaDxeLib.inf |
| ArmPkg/Library/ArmSmcLibNull/ArmSmcLibNull.inf | MdePkg/Library/ArmSmcLibNull/ArmSmcLibNull.inf |

Signed-off-by: Kun Qin <kun.qin@microsoft.com>

Raymond-MS reviewed (approved):
LGTM!
